drivers/adc/ad9083/ad9083.c

```diff
@@ -0,0 +1,385 @@
+		       uint8_t *readval)
+{
+	int32_t ret = 0;
+	uint8_t data[SPI_IN_OUT_BUFF_SZ] = {0};
+
+	if (reg < 0x1000) {
+		data[0] = (((reg >> 8) | 0x80) & 0xFF);
+		data[1] = (reg & 0xFF);
+		ret = spi_write_and_read(device->spi_desc, data, SPI_IN_OUT_BUFF_SZ);
+		if (ret != 0)
+			return ret;
+
+		*readval = data[2];
+	}
+	return SUCCESS;
+}
+
+/**
+ * @brief Write register.
+ * @param device - ad9083 device.
+ * @param reg - Address of the register.
+ * @param writeval - Value of the register.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t ad9083_reg_set(struct ad9083_phy *device, uint32_t reg,
+		       uint8_t writeval)
+{
+	int32_t ret = 0;
+	uint8_t data[SPI_IN_OUT_BUFF_SZ] = {0};
+
+	if (reg < 0x1000) {
+		data[0] = ((reg >> 8) & 0xFF);
+		data[1] = (reg & 0xFF);
+		data[2] = writeval;
+		ret = spi_write_and_read(device->spi_desc, data, SPI_IN_OUT_BUFF_SZ);
+		if (ret != 0)
+			return ret;
+	}
+	return SUCCESS;
+}
+
+/**
+ * Spi write and read compatible with ad9083 API
+ * @param user_data
+ * @param in_data - Pointer to array with the data to be sent on the SPI
+ * @param out_data - Pointer to array where the data to which the SPI will be written
+ * @param size_bytes - The size in bytes allocated for each of the indata and outdata arrays.
+ * @return 0 for success, any non-zero value indicates an error
+ */
+static int32_t ad9083_spi_xfer(void *user_data, uint8_t *in_data,
+			       uint8_t *out_data, uint32_t size_bytes)
+{
+	struct ad9083_phy *phy = user_data;
+	uint8_t data[6];
+	uint16_t bytes_number;
+	int32_t ret;
+	int32_t i;
+
+	bytes_number = (size_bytes & 0xFF);
+
+	for (i = 0; i < bytes_number; i++)
+		data[i] =  in_data[i];
+
+	ret = spi_write_and_read(phy->spi_desc, data, bytes_number);
+	if (ret != SUCCESS)
+		return FAILURE;
+
+	for (i = 0; i < bytes_number; i++)
+		out_data[i] =  data[i];
+
+	return SUCCESS;
+}
+
+/**
+ * Log write
+ * @param user_data
+ * @param log_type - Log source type.
+ * @param message - Message.
+ * @param argp - Argp.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t ad9083_log_write(void *user_data, int32_t log_type, const char *message,
+			 va_list argp)
+{
+	char logMessage[160];
+
+	vsnprintf(logMessage, sizeof(logMessage), message, argp);
+
+	switch (log_type) {
+	case ADI_CMS_LOG_NONE:
+		break;
+	case ADI_CMS_LOG_MSG:
+		break;
+	case ADI_CMS_LOG_WARN:
+		printf("%s\n", logMessage);
+		break;
+	case ADI_CMS_LOG_ERR:
+		printf("%s\n", logMessage);
+		break;
+	case ADI_CMS_LOG_SPI:
+		break;
+	case ADI_CMS_LOG_API:
+		break;
+	case ADI_CMS_LOG_ALL:
+		printf(logMessage);
+		break;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * Delay microseconds, compatible with ad9083 API
+ * Performs a blocking or sleep delay for the specified time in microseconds
+ * @param user_data
+ * @param us - time to delay/sleep in microseconds.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static int ad9083_udelay(void *user_data, unsigned int us)
```

Comment:
is this function needed? Can't you use directly `udelay()`?

Reply:
typedef int32_t(*adi_delay_us_t)(void *user_data, uint32_t us);